presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeToTimestampCast.java

```diff
@@ -0,0 +1,96 @@
+import static io.prestosql.type.DateTimes.PICOSECONDS_PER_SECOND;
+import static io.prestosql.type.DateTimes.SECONDS_PER_DAY;
+import static io.prestosql.type.DateTimes.rescale;
+import static io.prestosql.type.DateTimes.round;
+import static io.prestosql.type.DateTimes.scaleEpochMicrosToMillis;
+import static java.lang.Math.multiplyExact;
+
+@ScalarOperator(CAST)
+public final class TimeToTimestampCast
+{
+    private TimeToTimestampCast() {}
+
+    @LiteralParameters({"sourcePrecision", "targetPrecision"})
+    @SqlType("timestamp(targetPrecision)")
+    public static long castToShort(
+            @LiteralParameter("sourcePrecision") long sourcePrecision,
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            ConnectorSession session,
+            @SqlType("time(sourcePrecision)") long time)
+    {
+        long epochMicros = computeEpochMicros(sourcePrecision, targetPrecision, session, time);
+        if (targetPrecision <= 3) {
+            return scaleEpochMicrosToMillis(epochMicros);
+        }
+
+        return epochMicros;
+    }
+
+    @LiteralParameters({"sourcePrecision", "targetPrecision"})
+    @SqlType("timestamp(targetPrecision)")
+    public static LongTimestamp castToLong(
+            @LiteralParameter("sourcePrecision") long sourcePrecision,
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            ConnectorSession session,
+            @SqlType("time(sourcePrecision)") long time)
+    {
+        long epochMicros = computeEpochMicros(sourcePrecision, targetPrecision, session, time);
+        return new LongTimestamp(epochMicros, (int) (time % PICOSECONDS_PER_MICROSECOND));
+    }
+
+    private static long computeEpochMicros(@LiteralParameter("sourcePrecision") long sourcePrecision, @LiteralParameter("targetPrecision") long targetPrecision, ConnectorSession session, @SqlType("time(sourcePrecision)") long time)
+    {
+        long epochDay = LocalDate.ofInstant(session.getStart(), session.getTimeZoneKey().getZoneId())
+                .toEpochDay();
+
+        long epochSeconds = multiplyExact(epochDay, SECONDS_PER_DAY) + time / PICOSECONDS_PER_SECOND;
+
+        if (session.isLegacyTimestamp()) {
+            Instant instant = Instant.ofEpochSecond(epochSeconds);
+            LocalDateTime dateTime = LocalDateTime.ofInstant(instant, session.getTimeZoneKey().getZoneId());
+            List<ZoneOffset> offsets = session.getTimeZoneKey().getZoneId().getRules().getValidOffsets(dateTime);
+            epochSeconds -= offsets.get(0).getTotalSeconds();
+        }
+
+        long picoFraction = time % PICOSECONDS_PER_SECOND;
+        if (sourcePrecision > targetPrecision) {
+            picoFraction = round(picoFraction, (int) (TimeType.MAX_PRECISION - targetPrecision));
+        }
+
+        // target precision is > 6 for long timestamp
+        long microFraction = rescale(picoFraction, TimeType.MAX_PRECISION, 6);
```

Comment:
I don't understand the comment. Especially since this method is called in to-short and to-long cases

Reply:
Leftover from before I extracted the method. I'll remove the comment.